### PR TITLE
feat: add pysparnn base class indexer

### PR DESCRIPTION
**Pull request by davidbp**
Adding [PysparNN](https://github.com/facebookresearch/pysparnn)  integration as a new indexer for sparse data.

### CODE REVIEW

File: indexers/vector/PysparnnIndexer/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+jina
+numpy
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/requirements.txt:2]: numpy is a `jina` core dependency, so there is no point on adding it here. also `jina` is added in the `dockerfile`

**Review by JoanFM — COMMENTED** [review]
This is the proper idea, we need to take care now on storing the indices to disk.

Maybe we can start by using `pickle` as first approach.

File: indexers/vector/PysparnnIndexer/__init__.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    def _build_advanced_index(self):
+        keys = []
+        indexed_vectors = []
+        import pysparnn.cluster_index as ci
+        for key, vector in self.index.items():
+            keys.append(key)
+            indexed_vectors.append(vector)
+        
+        self.mci = ci.MultiClusterIndex(scipy.sparse.vstack(indexed_vectors), keys)
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/__init__.py:27]: `num_indexes` seems like it should be used in this `constructor`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: indexers/vector/PysparnnIndexer/__init__.py [hunk 3]
```diff
@@ -104,19 +104,3 @@ def delete(self, keys, *args, **kwargs) -> None:
         """
         for key in keys:
             del self.index[key]
-
-    def store_index_to_disk(self):
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/__init__.py]: do not remove this, it was added as a pre step to know how to store data

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: indexers/vector/PysparnnIndexer/__init__.py [hunk 4]
```diff
@@ -50,7 +50,10 @@ def close(self) -> None:
         super().close()
 
     def _assign_distance_class(self, metric: str):
+        import pysparnn
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/__init__.py:53]: remove this import

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: indexers/vector/PysparnnIndexer/__init__.py [hunk 5]
```diff
@@ -27,20 +26,35 @@ def __init__(self,
...
+    def get_write_handler(self):
+        pass
+
+    def get_query_handler(self):
+        pass
+
     def post_init(self):
         import os
         super().post_init()
+        self.embeddings = []
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/__init__.py:51]: Remove `self.embeddings and `self.indices`. They are not used for anything. Leftovers from some tests

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py [hunk 6]
```diff
@@ -102,7 +102,10 @@ def test_add_query_add_with_closing(indexer, features):
     indexer_query = PysparnnIndexer(metas=indexer.metas, metric='cosine')
     #  ValueError: Not possible query while indexing
     indexer_query.add(keys=list(range(0, 50)), vectors=features)
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:104]: this `add` is not so much an add as it is an `update`. It is expected to add new "non-existing" keys

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py [hunk 7]
```diff
@@ -43,6 +43,12 @@ def test_add(indexer, features):
     assert isinstance(indexer.index[0], csr_matrix)
 
 
+def test_add_newkeys(indexer, features):
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:46]: This does not test adding new keys,every time the `indexer` fixture is created, a new indexer is created with no data.

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py [hunk 8]
```diff
@@ -119,3 +133,7 @@ def test_delete_close_load(indexer, features):
     indexer_index_values = [ indexer.index[k] for k in indexer_index_keys]
     indexer_from_file_values = [ indexer_from_file.index[k] for k in indexer_index_keys]
     assert  np.array([x.nnz==y.nnz for x,y in  zip(indexer_index_values,indexer_from_file_values)]).all()
+
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:136]: Watch, remove empty lines at the end

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py [hunk 9]
```diff
@@ -100,13 +106,21 @@ def test_add_query_add_with_closing(indexer, features):
...
     indexer_query = PysparnnIndexer(metas=indexer.metas, metric='cosine')
-    #  ValueError: Not possible query while indexing
     indexer_query.add(keys=list(range(0, 50)), vectors=features)
     indexer_query_results = indexer_query.query(vectors=features[:5], top_k=1)
 
     assert (query_results[0] == indexer_query_results[0]).all()    
     assert (query_results[1] == indexer_query_results[1]).all()
 
+
+def test_add_close_add_newkeys(indexer, features):
```
Comment (JoanFM) [inline indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:116]: can u add some examples like this that also leverage `update` and `delete` API?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py (hunk 8 above)
Comment (davidbp) [reply indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:136]: Done

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py (hunk 7 above)
Comment (davidbp) [reply indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:46]: Done

File: indexers/vector/PysparnnIndexer/__init__.py (hunk 5 above)
Comment (davidbp) [reply indexers/vector/PysparnnIndexer/__init__.py]: Done

File: indexers/vector/PysparnnIndexer/__init__.py (hunk 4 above)
Comment (davidbp) [reply indexers/vector/PysparnnIndexer/__init__.py]: Done

File: indexers/vector/PysparnnIndexer/__init__.py (hunk 2 above)
Comment (davidbp) [reply indexers/vector/PysparnnIndexer/__init__.py]: yes, done

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py (hunk 8 above)
Comment (JoanFM) [reply indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:136]: I still se 5 blank lines at the end

**Review by JoanFM — COMMENTED** [review]
Looks good. Let's make sure we have proper docstrings and README

File: indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py (hunk 9 above)
Comment (davidbp) [reply indexers/vector/PysparnnIndexer/tests/test_pysparnnindexer.py:121]: done

**Pull request merged**